### PR TITLE
[WIP] - Add fix-certs cmd

### PR DESCRIPTION
**Pull request by tnozicka**
Requires:
 - [ ] https://github.com/openshift/cluster-kube-apiserver-operator/pull/448

/cc @mfojtik @deads2k

### CODE REVIEW

**Comment by tnozicka** [comment]
recovery-apiserver got it's own PR https://github.com/openshift/cluster-kube-apiserver-operator/pull/448

**Comment by sttts** [comment]
Commit message is bad of second commit.

File: pkg/cmd/fixcerts/fixcerts.go [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+func NewCommand() *cobra.Command {
+	o := &Options{
+		KubeApiserverImage:    "", // TODO: set the public pullspec
+		PodManifestDir:        "/etc/kubernetes/manifests",
+		StaticPodResourcesDir: "/etc/kubernetes/static-pod-resources",
+		Timeout:               5 * time.Minute,
+	}
+
+	cmd := &cobra.Command{
+		Use: "fix-certs",
```
Comment (sttts) [inline pkg/cmd/fixcerts/fixcerts.go]: description is missing. What is this command doing?

File: pkg/cmd/fixcerts/fixcerts.go [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+	metav1.AddToGroupVersion(Scheme, schema.GroupVersion{Version: "v1"})
+	utilruntime.Must(corev1.AddToScheme(Scheme))
+}
+
+type Options struct {
+	// TODO: merge with CreateOptions
+	KubeApiserverImage    string
+	PodManifestDir        string
+	StaticPodResourcesDir string
+	Timeout               time.Duration
```
Comment (sttts) [inline pkg/cmd/fixcerts/fixcerts.go]: Timeout of what?

File: pkg/cmd/fixcerts/fixcerts.go [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+
+		SilenceErrors: true,
+		SilenceUsage:  true,
+	}
+
+	// TODO: add/prettify descriptions
+	cmd.Flags().StringVar(&o.KubeApiserverImage, "kube-apiserver-image", o.KubeApiserverImage, "")
+	cmd.Flags().StringVar(&o.StaticPodResourcesDir, "static-pod-resources", o.StaticPodResourcesDir, "path to store a directory containing static pod resources for recovery apiserver manifest")
+	cmd.Flags().StringVar(&o.PodManifestDir, "pod-manifest-dir", o.PodManifestDir, "directory for the static pod manifes")
+	cmd.Flags().DurationVar(&o.Timeout, "timeout", o.Timeout, "timeout, 0 means infinite")
```
Comment (sttts) [inline pkg/cmd/fixcerts/fixcerts.go]: Timeout of what?

**Comment by tnozicka** [comment]
yes, this one is in progress still

File: pkg/cmd/fixcerts/fixcerts.go [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+	ctx, cancel := watch.ContextWithOptionalTimeout(context.Background(), o.Timeout)
+	defer cancel()
+	// TODO: hook up signals
+
+	apiserver := &recovery.Apiserver{
+		KubeApiserverImage:    o.KubeApiserverImage,
+		PodManifestDir:        o.PodManifestDir,
+		StaticPodResourcesDir: o.StaticPodResourcesDir,
+	}
+	defer apiserver.Destroy()
```
Comment (sttts) [inline pkg/cmd/fixcerts/fixcerts.go]: shouldn't this be after creation?

File: pkg/cmd/fixcerts/fixcerts.go [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+	if err != nil {
+		return err
+	}
+
+	configClient, err := configeversionedclient.NewForConfig(restConfig)
+	if err != nil {
+		return fmt.Errorf("failed to create config client: %v", err)
+	}
+
+	operatorConfigInformers := operatorexternalinformers.NewSharedInformerFactory(operatorConfigClient, 10*time.Minute)
```
Comment (sttts) [inline pkg/cmd/fixcerts/fixcerts.go]: missing the start of both shared informer factories.

File: pkg/cmd/fixcerts/fixcerts.go [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+		kubeControllerManagerOperatorNamespace,
+		kubeControllerManagerNamespace,
+		"kube-system",
+	)
+
+	eventRecorder := events.NewKubeRecorder(kubeClient.CoreV1().Events(""), "fix-certs (CLI)", &corev1.ObjectReference{
+		APIVersion: "v1",
+		Kind:       "namespace",
+		Name:       "kube-system",
+	}) // fake
```
Comment (sttts) [inline pkg/cmd/fixcerts/fixcerts.go]: what is fake?

File: pkg/cmd/fixcerts/fixcerts.go [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+	if err != nil {
+		return err
+	}
+
+	certRotationCtx, certRotationCtxCancel := context.WithCancel(ctx)
+	defer certRotationCtxCancel()
+
+	certRotationWg := sync.WaitGroup{}
+	go func() {
+		certRotationWg.Add(1)
```
Comment (sttts) [inline pkg/cmd/fixcerts/fixcerts.go]: add outside of the loop. This can race.

File: pkg/cmd/fixcerts/fixcerts.go [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+	kubeApiserverPod, err := apiserver.GetCurrentApiserverManifest()
+	if err != nil {
+		return fmt.Errorf("failed to find kube-apiserver's resource-dir: %v", err)
+	}
+	kubeApiserverCertsDir, err := recovery.GetVolumeHostPathPath("cert-dir", kubeApiserverPod.Spec.Volumes)
+	if err != nil {
+		return fmt.Errorf("failed to find kube-apiserver certs dir: %v", err)
+	}
+
+	kubeControllerManagerManifest := path.Join(o.PodManifestDir, KubeControllerManagerStaticPodFileName)
```
Comment (sttts) [inline pkg/cmd/fixcerts/fixcerts.go]: filepath

**Comment by tnozicka** [comment]
@sttts this is WIP still, I'll let you know when it is ready to review, I have not even went through it myself yet :)

File: pkg/cmd/fixcerts/fixcerts.go [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+				return err
+			}
+			for k, v := range configMap.Data {
+				data[k] = []byte(v)
+			}
+		default:
+			return fmt.Errorf("unknown object type %q", def.objectType)
+		}
+
+		dir := path.Join(def.toplevelDir, def.objectType, def.name)
```
Comment (sttts) [inline pkg/cmd/fixcerts/fixcerts.go]: filepath

File: pkg/cmd/fixcerts/fixcerts.go [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+		err = os.MkdirAll(dir, 0755)
+		if err != nil {
+			return fmt.Errorf("failed to create dir %q: %v", dir, err)
+		}
+
+		for name, bytes := range data {
+			filePath := path.Join(dir, name)
+			err := recovery.EnsureFileContent(filePath, bytes)
```
Comment (sttts) [inline pkg/cmd/fixcerts/fixcerts.go]: does this use 0600 for secrets?

File: pkg/cmd/fixcerts/kubecontrollermanagercertrotation.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	"github.com/openshift/cluster-kube-controller-manager-operator/pkg/operator/operatorclient"
+)
+
+// defaultRotationDay is the default rotation base for all cert rotation operations.
+const defaultRotationDay = 24 * time.Hour
+
+type CertRotationController struct {
```
Comment (sttts) [inline pkg/cmd/fixcerts/kubecontrollermanagercertrotation.go]: godoc, what is this?

File: pkg/cmd/fixcerts/kubecontrollermanagercertrotation.go [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+	ret := &CertRotationController{}
+
+	rotationDay := defaultRotationDay
+	if day != time.Duration(0) {
+		rotationDay = day
+		klog.Warningf("!!! UNSUPPORTED VALUE SET !!!")
+		klog.Warningf("Certificate rotation base set to %q", rotationDay)
+	}
+
+	certRotator, err := certrotation.NewCertRotationController(
```
Comment (sttts) [inline pkg/cmd/fixcerts/kubecontrollermanagercertrotation.go]: is this duplicated code? From where?

File: Dockerfile [hunk 13]
```diff
@@ -11,3 +11,6 @@ COPY --from=builder /go/src/github.com/openshift/cluster-kube-apiserver-operator
 COPY manifests/*.yaml /manifests
 COPY manifests/image-references /manifests
 LABEL io.openshift.release.operator true
+# FIXME: entrypoint shouldn't be bash but the binary (needs fixing the chain)
```
Comment (mfojtik) [inline Dockerfile]: remove

File: Dockerfile.rhel7 [hunk 14]
```diff
@@ -11,3 +11,5 @@ COPY --from=builder /go/src/github.com/openshift/cluster-kube-apiserver-operator
 COPY manifests/*.yaml /manifests
 COPY manifests/image-references /manifests
 LABEL io.openshift.release.operator true
+# FIXME: entrypoint shouldn't be bash but the binary (needs fixing the chain)
```
Comment (mfojtik) [inline Dockerfile.rhel7]: remove

File: pkg/cmd/recoveryapiserver/create.go [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+	KubeApiserverImage    string
+	StaticPodResourcesDir string
+	Timeout               time.Duration
+	Wait                  bool
+}
+
+func NewCreateCommand() *cobra.Command {
+	o := &CreateOptions{
+		Options:               NewDefaultOptions(),
+		KubeApiserverImage:    "", // TODO: set the public pullspec
```
Comment (mfojtik) [inline pkg/cmd/recoveryapiserver/create.go]: remove, error out in Validate() ?

File: pkg/cmd/recoveryapiserver/create.go [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+
+	if o.StaticPodResourcesDir == "" {
+		return errors.New("missing required flag --static-pod-resources")
+	}
+
+	return nil
+}
+
+func (o *CreateOptions) Run() error {
+	ctx, cancel := watch.ContextWithOptionalTimeout(context.Background(), o.Timeout)
```
Comment (mfojtik) [inline pkg/cmd/recoveryapiserver/create.go]: context.TODO()

Comment (mfojtik) [reply pkg/cmd/recoveryapiserver/create.go]: in future we will pass the context via Run() that context will be wired to TERM signal handler

File: pkg/cmd/recoveryapiserver/create.go [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+	defer cancel()
+	// TODO: hook up signals
```
Comment (mfojtik) [inline pkg/cmd/recoveryapiserver/create.go]:
```
			termHandler := server.SetupSignalHandler()
			ctx, shutdown := context.WithCancel(context.TODO())
			go func() {
				defer shutdown()
				<-termHandler
			}()
```

File: pkg/cmd/recoveryapiserver/create.go [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+	}
+
+	kubeconfigPath := path.Join(apiserver.GetRecoveryDir(), recovery.AdminKubeconfigFileName)
+	fmt.Printf("system:admin kubeconfig written to %q\n", kubeconfigPath)
+
+	if !o.Wait {
+		return nil
+	}
+
+	fmt.Printf("Waiting for recovery apiserver to come up.")
```
Comment (mfojtik) [inline pkg/cmd/recoveryapiserver/create.go]: use klog or at least add new line ;-)

File: pkg/recovery/apiserver.go [hunk 19]
```diff
@@ -0,0 +1,393 @@
...
+	return s.restConfig, nil
+}
+
+func (s *Apiserver) KubeConfig() (*clientcmdapiv1.Config, error) {
+	restConfig, err := s.RestConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	return &clientcmdapiv1.Config{
```
Comment (mfojtik) [inline pkg/recovery/apiserver.go]: `rest.CopyConfig`

File: pkg/recovery/apiserver.go [hunk 20]
```diff
@@ -0,0 +1,393 @@
...
+				AuthInfo: clientcmdapiv1.AuthInfo{
+					ClientCertificateData: restConfig.CertData,
+					ClientKeyData:         restConfig.KeyData,
+				},
+			},
+		},
+	}, nil
+}
+
+func (s *Apiserver) GetKubeClientset() (*kubernetes.Clientset, error) {
```
Comment (mfojtik) [inline pkg/recovery/apiserver.go]: why you need clientset?

File: pkg/cmd/fixcerts/fixcerts.go [hunk 21]
```diff
@@ -0,0 +1,407 @@
...
+
+			return nil
+		},
+
+		SilenceErrors: true,
+		SilenceUsage:  true,
+	}
+
+	// TODO: add/prettify descriptions
+	cmd.Flags().StringVar(&o.KubeApiserverImage, "kube-apiserver-image", o.KubeApiserverImage, "")
```
Comment (deads2k) [inline pkg/cmd/fixcerts/fixcerts.go]: same as other pull, I don't want to choose images or directories.

File: pkg/cmd/fixcerts/fixcerts.go [hunk 22]
```diff
@@ -0,0 +1,407 @@
...
+
+	return nil
+}
+
+func (o *Options) Run() error {
+	ctx, cancel := watch.ContextWithOptionalTimeout(context.Background(), o.Timeout)
+	defer cancel()
+	// TODO: hook up signals
+
+	apiserver := &recovery.Apiserver{
```
Comment (deads2k) [inline pkg/cmd/fixcerts/fixcerts.go]: this looks weird.

File: pkg/cmd/fixcerts/fixcerts.go [hunk 23]
```diff
@@ -0,0 +1,407 @@
...
+		KubeApiserverImage:    o.KubeApiserverImage,
+		PodManifestDir:        o.PodManifestDir,
+		StaticPodResourcesDir: o.StaticPodResourcesDir,
+	}
+
+	err := apiserver.Create()
```
Comment (deads2k) [inline pkg/cmd/fixcerts/fixcerts.go]: We should not do this.  Fixing the certs should use the `admin.kubeconfig` from it's known location.  It should not try to start a kube-apiserver

File: pkg/cmd/fixcerts/fixcerts.go [hunk 24]
```diff
@@ -0,0 +1,407 @@
...
+	}
+
+	certRotationCtx, certRotationCtxCancel := context.WithCancel(ctx)
+	defer certRotationCtxCancel()
+
+	certRotationWg := sync.WaitGroup{}
+
+	operatorConfigInformers := operatorexternalinformers.NewSharedInformerFactory(operatorConfigClient, 10*time.Minute)
+	certRotationWg.Add(1)
+	go func() {
```
Comment (deads2k) [inline pkg/cmd/fixcerts/fixcerts.go]: This is an odd structure.  `Start` is non-blocking.  I don't see a reason to wrap it in a gofunc

File: pkg/cmd/fixcerts/fixcerts.go [hunk 25]
```diff
@@ -0,0 +1,431 @@
...
+		kubeapiserveroperatorclient.GlobalUserSpecifiedConfigNamespace,
+		kubeapiserveroperatorclient.GlobalMachineSpecifiedConfigNamespace,
+		kubeapiserveroperatorclient.TargetNamespace,
+		kubeapiserveroperatorclient.OperatorNamespace,
+		"kube-system",
+	)
+	certRotationWg.Add(1)
+	go func() {
+		defer certRotationWg.Done()
+		kubeApiserverInformersForNamespaces.Start(certRotationCtx.Done())
```
Comment (deads2k) [inline pkg/cmd/fixcerts/fixcerts.go]: same coment about gofunc

File: pkg/cmd/fixcerts/fixcerts.go [hunk 26]
```diff
@@ -0,0 +1,431 @@
...
+	certRotationWg.Add(1)
+	go func() {
+		defer certRotationWg.Done()
+		kubeControllerManagerInformersForNamespaces.Start(certRotationCtx.Done())
+	}()
+
+	eventRecorder := events.NewKubeRecorder(kubeClient.CoreV1().Events(""), "fix-certs (CLI)", &corev1.ObjectReference{
+		APIVersion: "v1",
+		Kind:       "namespace",
+		Name:       "kube-system",
```
Comment (deads2k) [inline pkg/cmd/fixcerts/fixcerts.go]: use the `openshift-kube-apiserver-operator` namespace and set the namespace value or we end up in default and unfindable.

File: pkg/cmd/fixcerts/fixcerts.go [hunk 27]
```diff
@@ -0,0 +1,431 @@
...
+	go func() {
+		defer certRotationWg.Done()
+
+		kubeControllerManagerCertRotationController.Run(1, certRotationCtx.Done())
+	}()
+
+	klog.Info("Waiting for certs to be refreshed...")
+	// FIXME: wait for valid certs
+	// time.Sleep(5*time.Minute)
+	time.Sleep(30 * time.Second)
```
Comment (deads2k) [inline pkg/cmd/fixcerts/fixcerts.go]: This won't do.  Add a `RunOnce` at each level so we can be sure we run and avoid the go func/defer, wait, hope flow.

Comment (deads2k) [reply pkg/cmd/fixcerts/fixcerts.go]: > This won't do. Add a `RunOnce` at each level so we can be sure we run and avoid the go func/defer, wait, hope flow.

You'll also need a `WaitForPrereqs` func to sync the bits at least once.

File: pkg/cmd/fixcerts/fixcerts.go [hunk 28]
```diff
@@ -0,0 +1,431 @@
...
+			filePath := filepath.Join(dir, name)
+			err := recovery.EnsureFileContent(filePath, bytes)
+			if err != nil {
+				return fmt.Errorf("failed to write file %q: %v", filePath, err)
+			}
+		}
+	}
+
+	timestamp := time.Now().Format(time.RFC3339)
+	kubeControllerManagerPod.Annotations["force-triggered-by-fix-certs-at"] = timestamp
```
Comment (deads2k) [inline pkg/cmd/fixcerts/fixcerts.go]: I expected to see the kube-apiserver pod here too.  Let's do them all, just to be sure.

File: Dockerfile (hunk 13 above)
Comment (tnozicka) [reply Dockerfile]: @mfojtik I'd expect a reason

File: pkg/recovery/apiserver.go (hunk 19 above)
Comment (tnozicka) [reply pkg/recovery/apiserver.go]: from where? you are also commenting on the wrong PR, this one is just picking up https://github.com/openshift/cluster-kube-apiserver-operator/pull/448

File: pkg/cmd/fixcerts/fixcerts.go (hunk 10 above)
Comment (tnozicka) [reply pkg/cmd/fixcerts/fixcerts.go]: keeps the perm of what was there

Comment (tnozicka) [reply pkg/cmd/fixcerts/fixcerts.go]: and fallback is 600

**Comment by deads2k** [comment]
merged in https://github.com/openshift/cluster-kube-apiserver-operator/pull/461

**Pull request closed without merging**